### PR TITLE
Add tree grid

### PR DESCRIPTION
**Pull request by korbinancell**
This should solve a few of the issues in https://github.com/Faithlife/styled-ui/issues/216
My plan is after this to go about porting everything to v5

### CODE REVIEW

File: components/grid/base-grid.jsx [hunk 1]
```diff
@@ -246,3 +241,34 @@ BaseTable.propTypes = {
 	handleGetRowId: PropTypes.func,
 	children: PropTypes.node,
 };
+
+function useParseChildrenSettings(children, additionalCellComponents = {}) {
```
Comment (bryanrsmith) [inline components/grid/base-grid.jsx]: This looks like a standard functions. I don't think it should be named like a hook.

File: components/grid/tree-grid.jsx [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+	const { gridApi, setGridApi, columnApi, setColumnApi } = useGridState();
+
+	// For styling the drop target rows
+	const previousHoveredRowNode = useRef();
+	const hoveredRowNode = useRef();
+	const draggedNode = useRef();
+	const dragDirection = useRef();
+
+	const heading = React.Children.toArray(children).find(child => child && child.type.isTreeGroup);
+	useEffect(() => {
```
Comment (bryanrsmith) [inline components/grid/tree-grid.jsx]: I think we could wrap this in a `if (process.env.NODE_ENV !== 'production')`

File: components/grid/tree-grid.jsx [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+	}, []); // eslint-disable-line react-hooks/exhaustive-deps
+
+	const getShouldShowDropTarget = useCallback(
+		onDirection => ({ rowIndex }) =>
+			hoveredRowNode.current &&
+			hoveredRowNode.current.rowIndex === rowIndex &&
+			draggedNode.current.rowIndex !== hoveredRowNode.current.rowIndex &&
+			dragDirection.current === onDirection &&
+			(!isValidDropTarget ||
+				isValidDropTarget(
```
Comment (bryanrsmith) [inline components/grid/tree-grid.jsx]: This is hard to read. Consider adding a body and using an `if` statement.

File: components/grid/tree-grid.jsx [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+
+TreeGrid.rowSelectionOptions = BaseGrid.rowSelectionOptions;
+
+TreeGrid.rowTypes = {
+	folder: 'folder',
+	item: 'item',
+};
+
+TreeGrid.expandedRowsOptions = {
+	// Why ag-grid?
```
Comment (bryanrsmith) [inline components/grid/tree-grid.jsx]: What does this mean?

File: components/grid/tree-grid.jsx [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+// depth first for now
+function getTreeWithoutId(id, tree) {
+	const filteredTree = tree.filter(item => item.id !== id);
+	if (filteredTree.length !== tree.length) {
+		return filteredTree;
+	}
+
+	for (const item of filteredTree) {
+		if (item.children && item.children.length) {
+			item.children = getTreeWithoutId(id, item.children);
```
Comment (bryanrsmith) [inline components/grid/tree-grid.jsx]: Mutating items in this collections seems bad.

File: components/grid/tree-grid.jsx [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+			return [data, ...tree];
+		}
+		newTree.splice(direction === dragDirections.up ? parentIndex : parentIndex + 1, 0, data);
+
+		return newTree;
+	}
+
+	for (const item of newTree) {
+		if (item.children) {
+			item.children = insertDataIntoTree(data, parentId, direction, item.children);
```
Comment (bryanrsmith) [inline components/grid/tree-grid.jsx]: Here, too.

File: components/grid/tree-grid.jsx [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+
+	return newTree;
+}
+
+function getNewPath(parentNode, direction) {
+	const isParentFolder = parentNode.data.treeNodeType === TreeGrid.rowTypes.folder;
+	if (isParentFolder && parentNode.expanded && direction === dragDirections.down) {
+		return [...parentNode.data.path];
+	} else {
+		return [...parentNode.data.path.slice(0, -1)];
```
Comment (bryanrsmith) [inline components/grid/tree-grid.jsx]: `Array.prototype.slice` creates a shallow copy of the subsection of the array. You don't need to copy it again.

File: components/grid/tree-grid.jsx (hunk 4 above)
Comment (korbinancell) [reply components/grid/tree-grid.jsx]: It was a comment to self that slipped through. The different options are weird.

**Pull request merged**